src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs

```diff
@@ -1552,34 +1552,55 @@ public static SyntaxTree SyntaxTree(SyntaxNode root, ParseOptions? options = nul
         }
 
 #pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
-        /// <summary>
-        /// Produces a syntax tree by parsing the source text.
-        /// </summary>
+
+        /// <inheritdoc cref="CSharpSyntaxTree.ParseText(string, CSharpParseOptions?, string, Encoding?, CancellationToken)"/>
         public static SyntaxTree ParseSyntaxTree(
             string text,
             ParseOptions? options = null,
             string path = "",
             Encoding? encoding = null,
-            ImmutableDictionary<string, ReportDiagnostic>? diagnosticOptions = null,
-            bool? isGeneratedCode = null,
             CancellationToken cancellationToken = default)
+        {
+            return CSharpSyntaxTree.ParseText(text, (CSharpParseOptions?)options, path, encoding, cancellationToken);
+        }
+
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [Obsolete("The diagnosticOptions and isGeneratedCode parameters are obsolete due to performance problems, if you are using them use CompilationOptions.SyntaxTreeOptionsProvider instead")]
+        public static SyntaxTree ParseSyntaxTree(
+            string text,
+            ParseOptions? options,
+            string path,
+            Encoding? encoding,
+            ImmutableDictionary<string, ReportDiagnostic>? diagnosticOptions,
+            bool? isGeneratedCode,
+            CancellationToken cancellationToken)
         {
             return ParseSyntaxTree(SourceText.From(text, encoding), options, path, diagnosticOptions, isGeneratedCode, cancellationToken);
         }
 
-        /// <summary>
-        /// Produces a syntax tree by parsing the source text.
-        /// </summary>
+        /// <inheritdoc cref="CSharpSyntaxTree.ParseText(SourceText, CSharpParseOptions?, string, CancellationToken)"/>
         public static SyntaxTree ParseSyntaxTree(
             SourceText text,
             ParseOptions? options = null,
             string path = "",
-            ImmutableDictionary<string, ReportDiagnostic>? diagnosticOptions = null,
-            bool? isGeneratedCode = null,
             CancellationToken cancellationToken = default)
+        {
+            return CSharpSyntaxTree.ParseText(text, (CSharpParseOptions?)options, path, cancellationToken);
+        }
+
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [Obsolete("The diagnosticOptions and isGeneratedCode parameters are obsolete due to performance problems, if you are using them use CompilationOptions.SyntaxTreeOptionsProvider instead")]
+        public static SyntaxTree ParseSyntaxTree(
```

Comment:
Our general strategy with these is to move them to the bottom of the file.

Reply:
I moved it back up because it was making it really hard to maintain. Having the two main overloads in the middle of the file and then the rest in a scattered order was resulting in so much scrolling I ended up moving it for sanity.

Reply:
And also this is funky since the obsolete one is still also the implementation. :-/

Reply:
I'd prefer to keep with the existing precedent for where things are found. It's not just this file: all files where we deprecate APIs this in the compiler codebase do this.

---
In reply to: [460242888](https://github.com/dotnet/roslyn/pull/44331#discussion_r460242888) [](ancestors = 460242888)

Reply:
Ah so I might have confused the files -- I moved these back down; the ones in SyntaxTree where the implementation is obsolete but that's where the actual code lives I'm leaving as is.